File: blue-core/src/main/java/blue/orchestra/JavaScriptInstrument.java
package blue.orchestra;

import blue.Tables;
import blue.plugin.InstrumentPlugin;
import blue.scripting.JavaScriptProxy;
import blue.udo.OpcodeList;
import blue.utility.TextUtilities;
import blue.utility.UDOUtilities;
import electric.xml.Element;
import electric.xml.Elements;
import java.util.HashMap;
import javax.script.ScriptException;
import org.openide.util.Exceptions;

/**
 * <p>
 * Title: blue
 * </p>
 * <p>
 * Description: an object composition environment for csound
 * </p>
 * <p>
 * Copyright: Copyright (c) 2001-2002
 * </p>
 * <p>
 * Company: steven yi music
 * </p>
 * 
 * @author unascribed
 * @version 1.0
 */

@InstrumentPlugin(displayName = "JavaScriptInstrument", position = 30)
public class JavaScriptInstrument extends AbstractInstrument {

    String instrumentText;

    String globalOrc, globalSco;

    OpcodeList opcodeList;

    private transient HashMap udoReplacementValues;

    public JavaScriptInstrument() {
        instrumentText = "//use variable instrument at end of script to "
                + "bring instrument back into blue\n\n"
                + "instrument = \"aout oscili 32000, 440, 1\";";
        globalOrc = "";
        globalSco = "";

        opcodeList = new OpcodeList();
    }

    /** Copy Constructor */
    public JavaScriptInstrument(JavaScriptInstrument jsInstr) {
        super(jsInstr);
        instrumentText = jsInstr.instrumentText; 
        globalOrc = jsInstr.globalOrc; 
        globalSco = jsInstr.globalSco; 
        opcodeList = new OpcodeList(jsInstr.opcodeList); 
    }

    public String getText() {
        return instrumentText;
    }

    public void setText(String iText) {
        this.instrumentText = iText;
    }

    public boolean hasFTable() {
        return false;
    }

    @Override
    public void generateUserDefinedOpcodes(OpcodeList udoList) {
        udoReplacementValues = UDOUtilities.appendUserDefinedOpcodes(
                opcodeList, udoList);
    }

    @Override
    public void generateFTables(Tables tables) {
    }

    // -------------------------------------------

    @Override
    public String generateInstrument() {
        try {
            String retVal = JavaScriptProxy.processJavascriptInstrument(this.getText(),
                    this.getName());
            
            if (udoReplacementValues != null) {
                retVal = TextUtilities.replaceOpcodeNames(udoReplacementValues,
                        retVal);
                udoReplacementValues = null;
            }
            
            return retVal;
        } catch (ScriptException ex) {
            Exceptions.printStackTrace(ex);
            return "";
        }
    }

    @Override
    public String toString() {
        return this.name;
    }

    /*
     * (non-Javadoc)
     * 
     * @see blue.orchestra.Instrument#generateGlobalOrc()
     */
    @Override
    public String generateGlobalOrc() {
        return getGlobalOrc();
    }

    /*
     * (non-Javadoc)
     * 
     * @see blue.orchestra.Instrument#generateGlobalSco()
     */
    @Override
    public String generateGlobalSco() {
        return getGlobalSco();
    }

    /**
     * @return Returns the globalOrc.
     */
    public String getGlobalOrc() {
        return globalOrc;
    }

    /**
     * @param globalOrc
     *            The globalOrc to set.
     */
    public void setGlobalOrc(String globalOrc) {
        this.globalOrc = globalOrc;
    }

    /**
     * @return Returns the globalSco.
     */
    public String getGlobalSco() {
        return globalSco;
    }

    /**
     * @param globalSco
     *            The globalSco to set.
     */
    public void setGlobalSco(String globalSco) {
        this.globalSco = globalSco;
    }

    public static Instrument loadFromXML(Element data) throws Exception {
        JavaScriptInstrument instr = new JavaScriptInstrument();

        InstrumentUtilities.initBasicFromXML(data, instr);

        Elements elements = data.getElements();

        while (elements.hasMoreElements()) {
            Element node = elements.next();
            String nodeName = node.getName();
            switch (nodeName) {
                case "globalOrc":
                    instr.setGlobalOrc(node.getTextString());
                    break;
                case "globalSco":
                    instr.setGlobalSco(node.getTextString());
                    break;
                case "instrumentText":
                    instr.setText(node.getTextString());
                    break;
                case "opcodeList":
                    instr.opcodeList = OpcodeList.loadFromXML(node);
                    break;
            }
        }

        return instr;
    }

    /*
     * (non-Javadoc)
     * 
     * @see blue.orchestra.Instrument#saveAsXML()
     */
    @Override
    public Element saveAsXML() {
        Element retVal = InstrumentUtilities.getBasicXML(this);

        retVal.addElement("globalOrc").setText(this.getGlobalOrc());
        retVal.addElement("globalSco").setText(this.getGlobalSco());
        retVal.addElement("instrumentText").setText(this.getText());

        retVal.addElement(opcodeList.saveAsXML());

        return retVal;
    }

    public OpcodeList getOpcodeList() {
        return opcodeList;
    }

    @Override
    public JavaScriptInstrument deepCopy() {
        return new JavaScriptInstrument(this);
    }
}

File: packages/blue-data/src/instruments/javascript-instrument.ts
import { Element } from '../serialization/xml-reader';
import { OpcodeList } from '../opcodes/opcode-list';
import { appendUserDefinedOpcodes } from '../opcodes/udo-utilities';
import { Instrument } from './instrument';

export class JavaScriptInstrument extends Instrument {
  private _text =
    '//use variable instrument at end of script to bring instrument back into blue\n\n' +
    'instrument = "aout oscili 32000, 440, 1";';
  private _globalOrc = '';
  private _globalSco = '';
  private _opcodeList = new OpcodeList();

  constructor(other?: JavaScriptInstrument) {
    super();
    this.setName('JavaScriptInstrument');
    if (other) {
      this._name = other._name;
      this._enabled = other._enabled;
      this._comment = other._comment;
      this._text = other._text;
      this._globalOrc = other._globalOrc;
      this._globalSco = other._globalSco;
      this._opcodeList = OpcodeList.loadFromXML(other._opcodeList.saveAsXML());
    }
  }

  getText(): string {
    return this._text;
  }

  setText(text: string): void {
    this._text = text ?? '';
  }

  getGlobalOrc(): string {
    return this._globalOrc;
  }

  setGlobalOrc(globalOrc: string): void {
    this._globalOrc = globalOrc ?? '';
  }

  getGlobalSco(): string {
    return this._globalSco;
  }

  setGlobalSco(globalSco: string): void {
    this._globalSco = globalSco ?? '';
  }

  getOpcodeList(): OpcodeList {
    return this._opcodeList;
  }

  override generateUserDefinedOpcodes(udoList: unknown): void {
    if (!(udoList instanceof OpcodeList)) {
      return;
    }

    appendUserDefinedOpcodes(this._opcodeList, udoList);
  }

  override generateGlobalOrc(): string | null {
    return this._globalOrc || null;
  }

  override generateGlobalSco(): string | null {
    return this._globalSco || null;
  }

  override generateInstrument(): string {
    return '';
  }

  saveAsXML(): Element {
    const elem = new Element('instrument');
    elem.setAttribute('type', 'blue.orchestra.JavaScriptInstrument');
    elem.addElement('name').setText(this._name);
    elem.addElement('comment').setText(this._comment);
    elem.addElement('globalOrc').setText(this._globalOrc);
    elem.addElement('globalSco').setText(this._globalSco);
    elem.addElement('instrumentText').setText(this._text);
    elem.addElement(this._opcodeList.saveAsXML());
    return elem;
  }

  static loadFromXML(data: Element): JavaScriptInstrument {
    const instr = new JavaScriptInstrument();
    instr.setName(data.getTextString('name') ?? '');
    instr.setComment(data.getTextString('comment') ?? '');
    instr.setGlobalOrc(data.getTextString('globalOrc') ?? '');
    instr.setGlobalSco(data.getTextString('globalSco') ?? '');
    instr.setText(data.getTextString('instrumentText') ?? '');
    const opcodeList = data.getElement('opcodeList');
    if (opcodeList) instr._opcodeList = OpcodeList.loadFromXML(opcodeList);
    return instr;
  }

  deepCopy(): JavaScriptInstrument {
    return new JavaScriptInstrument(this);
  }
}
